src/components/protocol_handler/src/protocol_handler_impl.cc

```diff
@@ -488,20 +492,36 @@ void ProtocolHandlerImpl::SendStartSessionNAck(
 
   uint8_t maxProtocolVersion = SupportedSDLProtocolVersion();
 
+  utils::SemanticVersion full_version;
+  session_observer_.ProtocolVersionUsed(
+      connection_id, session_id, full_version);
+  const utils::SemanticVersion& min_version =
+      (full_version.major_version_ < PROTOCOL_VERSION_5)
+          ? default_protocol_version
+          : ((full_version < default_protocol_version)
+                 ? full_version
+                 : default_protocol_version);
+
   if (protocol_version >= PROTOCOL_VERSION_5 &&
-      maxProtocolVersion >= PROTOCOL_VERSION_5 && rejectedParams.size() > 0) {
+      maxProtocolVersion >= PROTOCOL_VERSION_5) {
     BsonObject payloadObj;
     bson_object_initialize_default(&payloadObj);
-    BsonArray rejectedParamsArr;
-    bson_array_initialize(&rejectedParamsArr, rejectedParams.size());
-    for (std::string param : rejectedParams) {
-      char paramPtr[256];
-      strncpy(paramPtr, param.c_str(), sizeof(paramPtr));
-      paramPtr[sizeof(paramPtr) - 1] = '\0';
-      bson_array_add_string(&rejectedParamsArr, paramPtr);
+    if (rejectedParams.size() > 0) {
+      BsonArray rejectedParamsArr;
+      bson_array_initialize(&rejectedParamsArr, rejectedParams.size());
+      for (std::string param : rejectedParams) {
+        char paramPtr[256];
+        strncpy(paramPtr, param.c_str(), sizeof(paramPtr));
+        paramPtr[sizeof(paramPtr) - 1] = '\0';
+        bson_array_add_string(&rejectedParamsArr, paramPtr);
+      }
+      bson_object_put_array(
+          &payloadObj, strings::rejected_params, &rejectedParamsArr);
+    }
+    if (!reason.empty() && min_version >= min_reason_param_version) {
```

Comment:
I wanted to mirror the same approach which is used by `StartSessionAck` https://github.com/smartdevicelink/sdl_core/blob/master/src/components/protocol_handler/src/protocol_handler_impl.cc#L313 but it looks like `StartSessionAck` has a few more major version checks which will prevent the bson payload from being added. 

I've added changes in https://github.com/smartdevicelink/sdl_core/pull/3462/commits/e8fa2afa899678b43c2bb64e51f9abcd5114ee19  to use the `full_version` instead of calculating the `min_version`

Reply:
Do you know the reason for this min version logic though? Maybe it is intended behavior that we use the latest protocol version if they don't report a valid version, and this line should use min_version.